in-progress/6973-refactor-base-contract-interaction.md

```diff
@@ -0,0 +1,1092 @@
+
+```ts
+const addContractClassRegistration = (
+  builder: TxExecutionRequestBuilder, call: UserFunctionCall
+) => {
+  if (!call.contractArtifact) {
+    throw new Error('Contract artifact must be provided to register class');
+  }
+
+  const contractClass = getContractClassFromArtifact(call.contractArtifact);
+
+  builder.addCapsule(
+    bufferAsFields(
+      contractClass.packedBytecode,
+      MAX_PACKED_PUBLIC_BYTECODE_SIZE_IN_FIELDS
+  ));
+
+  const { artifact, instance } = getCanonicalClassRegisterer();
+
+  const registerFnAbi = findFunctionAbi(artifact, 'register');
+
+  builder.addAppFunctionCall(
+    makeFunctionCall(
+      registerFnAbi,
+      instance.address,
+      {
+        artifact_hash: contractClass.artifactHash,
+        private_functions_root: contractClass.privateFunctionsRoot,
+        public_bytecode_commitment: contractClass.publicBytecodeCommitment
+      }
+  ));
+}
+```
+
+#### public deployment
+
+Define a helper somewhere as
+
+```ts
+
+const addPublicContractDeployment = (
+  builder: TxExecutionRequestBuilder, call: UserFunctionCall
+) => {
+  const { artifact, instance } = getCanonicalInstanceDeployer();
+  const deployFnAbi = findFunctionAbi(artifact, 'deploy');
+  builder.addAppFunctionCall(
+    makeFunctionCall(
+      deployFnAbi,
+      instance.address,
+      {
+        salt: request.contractInstance.salt,
+        contract_class_id: request.contractInstance.contractClassId,
+        initialization_hash: request.contractInstance.initializationHash,
+        public_keys_hash: request.contractInstance.publicKeysHash,
+        universal_deploy: request.contractInstance.deployer.isZero(),
+      }
+  ));
+}
+
+```
+
+#### Entrypoints implement `TxExecutionRequestComponent`
+
+Below is what a default account entrypoint might look like, but other entrypoints could implement this differently
+
+```ts
+
+export class DefaultAccountEntrypoint implements TxExecutionRequestComponent {
+  constructor(
+    private address: AztecAddress,
+    private auth: AuthWitnessProvider,
+    private chainId: number = DEFAULT_CHAIN_ID,
+    private version: number = DEFAULT_VERSION,
+  ) {}
+
+  async adaptTxExecutionRequest(
+    builder: TxExecutionRequestBuilder,
+    userRequest: UserRequest
+  ): Promise<void> {
+    const appPayload = EntrypointPayload.fromFunctionCalls(builder.appFunctionCalls);
+    const setupPayload = EntrypointPayload.fromFunctionCalls(builder.setupFunctionCalls);
+    const abi = this.getEntrypointAbi();
+    const entrypointPackedArgs = PackedValues.fromValues(encodeArguments(abi, [appPayload, setupPayload]));
+
+    return builder
+      .setOrigin(this.address)
+      .setTxContext({
+        chainId: this.chainId,
+        version: this.version,
+        gasSettings: userRequest.gasSettings,
+      })
+      .setFunctionSelector(
+        FunctionSelector.fromNameAndParameters(abi.name, abi.parameters)
+      )
+      .setFirstCallArgsHash(entrypointPackedArgs.hash());
```

Comment:
The `firstCallArgsHash` is not super telling to me 🤷 Think it should either be clearer from the name, or do the maker way of something that you don't "assume" what does.

Reply:
Totally agree. I think it should be called `entrypointPackedArgsHash`.